sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala

```diff
@@ -278,21 +278,48 @@ private[sql] object QueryExecutionErrors extends QueryErrorsBase with ExecutionE
   }
 
   def ansiDateTimeError(e: Exception): SparkDateTimeException = {
+    def extractDateTimeErrorInfo(e: Exception): (String, String, String) = {
+      val errorMessage = e.getMessage
+
+      val valuePattern = "Invalid value for ([A-Za-z]+) \\(valid values (.+)\\): (.+)".r
+      val datePattern = "Invalid date '[A-Z]+ ([0-9]+)'".r
+
+      errorMessage match {
+        case valuePattern(field, range, badValue) =>
+          val unit = field match {
+            case "Year" => "YEAR"
+            case "MonthOfYear" => "MONTH"
+            case "DayOfMonth" => "DAY"
+            case "HourOfDay" => "HOUR"
+            case "MinuteOfHour" => "MINUTE"
+            case "SecondOfMinute" => "SECOND"
+          }
+          val formattedRange = range.replace(" - ", " ... ")
+          (unit, formattedRange, badValue)
+        case datePattern(badDate) =>
+          ("DAY", "1 ... 28/31", badDate)
+      }
+    }
+    val (unit, range, badValue) = extractDateTimeErrorInfo(e)
     new SparkDateTimeException(
-      errorClass = "_LEGACY_ERROR_TEMP_2000",
+      errorClass = "DATE_TIME_FIELD_OUT_OF_BOUNDS",
       messageParameters = Map(
-        "message" -> e.getMessage,
-        "ansiConfig" -> toSQLConf(SQLConf.ANSI_ENABLED.key)),
+        "ansiConfig" -> toSQLConf(SQLConf.ANSI_ENABLED.key),
+        "unit" -> unit,
+        "range" -> range,
+        "badValue" -> badValue
+      ),
       context = Array.empty,
-      summary = "")
+      summary = "",
+      cause = Some(e)
+    )
   }
 
-  def ansiIllegalArgumentError(message: String): SparkIllegalArgumentException = {
-    new SparkIllegalArgumentException(
-      errorClass = "_LEGACY_ERROR_TEMP_2000",
-      messageParameters = Map(
-        "message" -> message,
-        "ansiConfig" -> toSQLConf(SQLConf.ANSI_ENABLED.key)))
+  def ansiIllegalArgumentError(): SparkException = {
+    SparkException.internalError(
```

Comment:
Are you sure that it is an internal error? What happens if ANSI on and `microseconds ! =0`:
```scala
  override def nullSafeEval(start: Any, interval: Any): Any = {
    val itvl = interval.asInstanceOf[CalendarInterval]
    if (ansiEnabled || itvl.microseconds == 0) {
      DateTimeUtils.dateAddInterval(start.asInstanceOf[Int], itvl)
    } else {
```
and user might see the error, correct?
```scala
  def dateAddInterval(
     start: Int,
     interval: CalendarInterval): Int = {
    if (interval.microseconds != 0) {
      throw QueryExecutionErrors.ansiIllegalArgumentError()
    }
```

Reply:
I see.

Then let me leave this it as a `LEGACY` error for now because this PR aims to provide proper error message for `make_date`/`make_timestamp`, but seems like this code path cannot be reachable via `make_date`/`make_timestamp`.